pypeit/coadd3d.py

```diff
@@ -0,0 +1,1772 @@
+        rel_humidity = self.spec.get_meta_value([hdr0], 'humidity')
+        coord = SkyCoord(raval, decval, unit=(units.deg, units.deg))
+        location = self.spec.location  # TODO :: spec.location should probably end up in the TelescopePar (spec.telescope.location)
+        # Set a default value
+        ra_corr, dec_corr = 0.0, 0.0
+        if pressure == 0.0:
+            msgs.warn("Pressure is set to zero - DAR correction will not be performed")
+        else:
+            msgs.info("DAR correction parameters:" + msgs.newline() +
+                      "   Pressure = {0:f} bar".format(pressure) + msgs.newline() +
+                      "   Temperature = {0:f} deg C".format(temperature) + msgs.newline() +
+                      "   Humidity = {0:f}".format(rel_humidity))
+            ra_corr, dec_corr = datacube.correct_dar(waves, coord, obstime, location,
+                                                     pressure * units.bar, temperature * units.deg_C, rel_humidity,
+                                                     wave_ref=wave_ref)
+        return ra_corr*cosdec, dec_corr
+
+    def align_user_offsets(self):
+        """
+        Align the RA and DEC of all input frames, and then
+        manually shift the cubes based on user-provided offsets.
+        The offsets should be specified in arcseconds, and the
+        ra_offset should include the cos(dec) factor.
+        """
+        # First, translate all coordinates to the coordinates of the first frame
+        # Note: You do not need cos(dec) here, this just overrides the IFU coordinate centre of each frame
+        #       The cos(dec) factor should be input by the user, and should be included in the self.opts['ra_offset']
+        ref_shift_ra = self.ifu_ra[0] - self.ifu_ra
+        ref_shift_dec = self.ifu_dec[0] - self.ifu_dec
+        for ff in range(self.numfiles):
+            # Apply the shift
+            self.all_ra[self.all_idx == ff] += ref_shift_ra[ff] + self.opts['ra_offset'][ff] / 3600.0
+            self.all_dec[self.all_idx == ff] += ref_shift_dec[ff] + self.opts['dec_offset'][ff] / 3600.0
+            msgs.info("Spatial shift of cube #{0:d}:" + msgs.newline() +
+                      "RA, DEC (arcsec) = {1:+0.3f} E, {2:+0.3f} N".format(ff + 1,
+                                                                           self.opts['ra_offset'][ff],
+                                                                           self.opts['dec_offset'][ff]))
+
+    def coadd(self):
+        """
+        Main entry routine to set the order of operations to coadd the data. For specific
+        details of this procedure, see the child routines.
+        """
+        msgs.bug("This routine should be overridden by child classes.")
+        msgs.error("Cannot proceed without coding the coadd routine.")
+        return
+
+
+class SlicerIFUCoAdd3D(CoAdd3D):
+    """
+    Child of CoAdd3D for SlicerIFU data reduction. For documentation, see CoAdd3d parent class above.
```

Comment:
Add some more documentation here indicating now a SlicerIFU coadd3d differs from the parent one.

Reply:
OK - more documentation added